### PR TITLE
Adding pika authentication and sleep time to queue consumer

### PR DESCRIPTION
**Pull request by lmarinve**
sleep time and pika authentication added to: sdx_controller/messaging/rpc_queue_consumer.py

### CODE REVIEW

File: sdx_controller/messaging/rpc_queue_consumer.py [hunk 1]
```diff
@@ -38,9 +46,17 @@ def on_request(self, ch, method, props, message_body):
...
+        self.logger = logging.getLogger(__name__)
+        SLEEP_TIME = 5
+        self.logger.info(" [*] Sleeping for %s seconds.", SLEEP_TIME)
+        time.sleep(SLEEP_TIME)
+
+        self.logger.info(" [*] Connecting to server ...")
+        credentials = pika.PlainCredentials(MQ_USER, MQ_PASS)
         self.connection = pika.BlockingConnection(
-            pika.ConnectionParameters(host=MQ_HOST)
+            pika.ConnectionParameters(MQ_SRVC, 5672, "/", credentials)
```
Comment (congwang09) [inline sdx_controller/messaging/rpc_queue_consumer.py:57]: It would be nice to have the MQ port number configurable, as an env var.

**Review by congwang09 — COMMENTED** [review]
Hey Luis, got some comments here.

File: sdx_controller/messaging/rpc_queue_consumer.py [hunk 2]
```diff
@@ -3,25 +3,33 @@
...
+MQ_SRVC = os.environ.get("MQ_SRVC")
+MQ_USER = os.environ.get("MQ_USER")
+MQ_PASS = os.environ.get("MQ_PASS")
 logger = logging.getLogger(__name__)
 
 
 class RpcConsumer(object):
     def __init__(self, thread_queue, exchange_name, topology_manager):
         self.logger = logging.getLogger(__name__)
+        SLEEP_TIME = 5
```
[the remark this answers is not in the document]
Comment (sajith) [reply sdx_controller/messaging/rpc_queue_consumer.py:25]: Why do we a need to add a delay anyway?

I suspect this is to wait for the RabbitMQ service to start up, in the case of RabbitMQ and SDX Controller spawned by the same compose file.  If that is the case, we should be using [depends_on](https://docs.docker.com/compose/compose-file/05-services/#depends_on) in the compose file.

File: Dockerfile [hunk 3]
```diff
@@ -1,16 +1,14 @@
...
 
+RUN mkdir -p /usr/src/app
 WORKDIR /usr/src/app
 
+<<<<<<< HEAD
+COPY ./container-sdx-controller /usr/src/app
+=======
 WORKDIR /usr/src/app
 COPY . /usr/src/app
+>>>>>>> main
```
Comment (sajith) [inline Dockerfile:11]: These changes to `Dockerfile` are not necessary, and they are not within the scope of this PR.  They contain errors too, with git merge markers. Please revert the changes.

File: sdx_controller/messaging/rpc_queue_consumer.py [hunk 4]
```diff
@@ -3,25 +3,33 @@
...
 import pika
 
 from sdx_controller.handlers.lc_message_handler import LcMessageHandler
 from sdx_controller.utils.parse_helper import ParseHelper
 
-MQ_HOST = os.environ.get("MQ_HOST")
 # subscribe to the corresponding queue
 SUB_QUEUE = os.environ.get("SUB_QUEUE")
-
+MQ_SRVC = os.environ.get("MQ_SRVC")
```
Comment (sajith) [inline sdx_controller/messaging/rpc_queue_consumer.py:16]: It looks like `MQ_HOST` was renamed to `MQ_SRVC`, but not with good reason. Everywhere else (environment files, tests, CI setup) we use `MQ_HOST`.

**Review by sajith — CHANGES_REQUESTED** [review]
The way I have understood, we can achieve password auth with this:

```python
MQ_USER = os.getenv("MQ_USER") or "guest"
MQ_PASS = os.getenv("MQ_PASS") or "guest"

creds = pika.PlainCredentials(username=MQ_USER, password=MQ_PASS)
params = pika.ConnectionParameters(host=MQ_HOST, credentials=cred)
```

That is the general idea.  Adding a sleep time is probably unnecessary, and if there is a reason, we should document it.

Also please be sure to run the tests in your local development setup.  How to do this is documented in the README file of this repository.  If something is unclear, we should address that.

File: sdx_controller/messaging/rpc_queue_consumer.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (lmarinve) [reply sdx_controller/messaging/rpc_queue_consumer.py:16]: Hi @congwang09, @sajith,  MQ_SRVC is a docker-compose variable that receives the existing .env variable RABBITMQ_DEFAULT_HOST, which has the docker-compose service name rabbitmq3

File: sdx_controller/messaging/rpc_queue_consumer.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (lmarinve) [reply sdx_controller/messaging/rpc_queue_consumer.py:25]: Hi @sajith, docker depends on is only responsible for managing the status of Docker services, not the internal services running inside the container, such as RabbitMQ. To avoid repetition, I plan to add a new variable to the .env file, as we will use this value multiple times in sdx lc and controller.

File: Dockerfile (hunk 3 above)
Comment (lmarinve) [reply Dockerfile:11]: There is a different Dockerfile that needs to be used while running in compose mode. To avoid affecting the original code, I moved the file to a different directory. There was an error on Github, and although I did a merge, the file was not updated. Therefore, I will update it manually.

File: sdx_controller/messaging/rpc_queue_consumer.py (hunk 4 above)
Comment (lmarinve) [reply sdx_controller/messaging/rpc_queue_consumer.py:16]: MQ_SRVC has the Docker service name value and MQ_HOST has the IP address. I suspect the test error is caused because adding authentication is not tested, and MQ_HOST is not used anymore.

File: Dockerfile (hunk 3 above)
Comment (congwang09) [reply Dockerfile:11]: This happens when there's conflict between your branch and your merged branch. You need to resolve the conflict between the branches.

File: sdx_controller/messaging/rpc_queue_consumer.py (hunk 4 above)
Comment (sajith) [reply sdx_controller/messaging/rpc_queue_consumer.py:16]: Please continue to use `MQ_HOST`, and please make just the necessary changes to make password authentication work.

File: sdx_controller/messaging/rpc_queue_consumer.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (sajith) [reply sdx_controller/messaging/rpc_queue_consumer.py:25]: I do not understand. The message queue would not be running in the same container.  What is the delay for?

File: sdx_controller/messaging/rpc_queue_consumer.py (hunk 4 above)
Comment (lmarinve) [reply sdx_controller/messaging/rpc_queue_consumer.py:16]: No problem! So, password authentication is currently working fine. I'll go ahead and bring back MQ_HOST, and if the test still doesn't go as planned, I'm happy to keep working on it. But I'd really appreciate your help in figuring out what's causing the issue and finding a solution to make the deployment happen. Thanks so much!

Comment (sajith) [reply sdx_controller/messaging/rpc_queue_consumer.py:16]: I'm looking into the test failures.

Comment (sajith) [reply sdx_controller/messaging/rpc_queue_consumer.py:16]: I think the problem is that not all environment variables are set when running the tests: especially `MQ_SRVC` (which should be `MQ_HOST`) in this case.  It will also fail if `MQ_USER` or `MQ_PASS` is not set.

It would be nice to have some checks to see that all configuration items are set, but that is a separate issue (see #27) from this.  We should definitely use some defaults in this PR -- like using RabbitMQ default username/password of `guest`/`guest` when no username and password environment variable is set, as I've stated in another review comment.

In the meanwhile, could you set up a local development environment and try running the tests yourself?

Comment (sajith) [reply sdx_controller/messaging/rpc_queue_consumer.py:16]: You probably want to reset your PR branch, and cherry-pick the relevant commits to the one relevant file, like so:

```
$ git switch luisdev
$ git fetch origin
$ git reset --hard origin/main
$ git cherry-pick 9ba4b92 c0d3459
```

Now, rename `MQ_SRVC` back to `MQ_HOST`, and use user/auth defaults.  Also remove the `time.sleep()` calls. We should not add unnecessary latency, unless there is a good reason.  And then update this PR with:

```
$ git push <remote> luisdev --force-with-lease
```

Obviously replace `<remote>` with whatever remote name that you are using.

**Comment by sajith** [comment]
Superseded by #223 and #250.  Let us close this now.
